ICSharpCode.Decompiler/CSharp/ProjectDecompiler/ProjectFileWriterSdkStyle.cs

```diff
@@ -0,0 +1,212 @@
+﻿// Copyright (c) 2020 Siegfried Pammer
+// 
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this
+// software and associated documentation files (the "Software"), to deal in the Software
+// without restriction, including without limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons
+// to whom the Software is furnished to do so, subject to the following conditions:
+// 
+// The above copyright notice and this permission notice shall be included in all copies or
+// substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+// PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+// FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection.PortableExecutable;
+using System.Xml;
+using ICSharpCode.Decompiler.Metadata;
+
+namespace ICSharpCode.Decompiler.CSharp.ProjectDecompiler
+{
+	/// <summary>
+	/// A <see cref="IProjectFileWriter"/> implementation that creates the projects in the SDK style format.
+	/// </summary>
+	sealed class ProjectFileWriterSdkStyle : IProjectFileWriter
+	{
+		const string AspNetCorePrefix = "Microsoft.AspNetCore";
+		const string PresentationFrameworkName = "PresentationFramework";
+		const string WindowsFormsName = "System.Windows.Forms";
+		const string TrueString = "True";
+		const string FalseString = "False";
+		const string AnyCpuString = "AnyCPU";
+
+		static readonly HashSet<string> ImplicitReferences = new HashSet<string> {
+			"mscorlib",
+			"netstandard",
+			"PresentationFramework",
+			"System",
+			"System.Diagnostics.Debug",
+			"System.Diagnostics.Tools",
+			"System.Drawing",
+			"System.Runtime",
+			"System.Runtime.Extensions",
+			"System.Windows.Forms",
+			"System.Xaml",
```

Comment:
This list is in no way exhaustive. I considered only those assemblies that will be referenced by a freshly built assembly created from a standard template (for WinForms, WPF, class lib and Web app).

The thing is, the SDK adds all needed references *automagically* depending on the usages in the user code. This list can become very large, see e.g. [this example](https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks.Packaging/src/FrameworkLists/.NETFramework%2CVersion%3Dv4.5/FrameworkList.xml). Furthermore, the lists differ for different targets (.NET Framework, .NET Core, ...) and are still changing for platforms in development (e.g. .NET 5).

What would you prefer?
- leave the list as it is
- keep only `mscorlib` and `netstandard`
- include as much as possible from the lists available today

Personally, I won't be satisfied with any of these 3, but we just can't re-implement the .NET SDK logic because, strictly speaking, it is not necessary that the target SDK is installed on the machine where decompiling with ILSpy occurs. So we can't ask the SDK itself about those reference assemblies...

Reply:
Leaving the list as-is seems like a good start. Just as decompilation cannot be perfect in all cases, we can expect the user to be able to fix the project file, if there are references missing.

@christophwille do you have any objections to this?

Reply:
No objections. Perfectly recreating the "original" csproj won't ever be possible anyways, and adding tons of guessing logic will only make the feature super hard to maintain.

Reply:
I am with you on that, @christophwille.

Just to clarify: the user might want to fix the resulting project file *by removing the unnecessary system references*. They won't harm, but they will be just redundant. The list above contains the assemblies that will be automatically de-referenced by ILSpy.

So in any case, the user should be able to build the project right away (even having redundant references). Of course provided that the source code can be built, which is not always the case (e.g. WPF projects already containing generated code originating from disassembled files: the SDK will generate stuff one more time causing conflicts.)